### PR TITLE
expression: replace recursive `Tree` datatype with a non-recursive one

### PR DESCRIPTION
**Pull request by apoelstra**
Replaces the recursive `expression::Tree` datatype with a nonrecursive one. Exposes a limited API consisting of operations which can be done efficiently (mainly, in-order iteration) and which are necessary for parsing trees and converting them to other types.

As a side effect this simplifies/unifies some more code and provides better error messages, in particular for threshold parsing. But that isn't a focus of this PR and I haven't quantified the changes.

This is the last of the "expression" PRs. I have followups which go in two directions: (1) eliminating more recursion and recursive datatypes, and (2) improving the TapTree API, which I found I needed this new expression API to do cleanly.

Will post benchmarks once I have them.

### CODE REVIEW

**Review by apoelstra — COMMENTED** [review]
On 515974a6ff6dfd03d8c60b391699b7678475806a successfully ran local tests

**Comment by apoelstra** [comment]
I don't understand this doc failure... it's claiming that the `Tree::verify_threshold` method doesn't exist but it definitely does (and I can call it from within the function I'm documenting).

**Review by apoelstra — COMMENTED** [review]
On 8235d5ffb669b6dae035dd04ebc9191b7b55fcc2 successfully ran local tests

**Comment by apoelstra** [comment]
Gentle ping @sanket1729

**Review by sanket1729 — APPROVED** [review]
ACK 8235d5ffb669b6dae035dd04ebc9191b7b55fcc2

Sorry for the delay in reviewing this PR. I wanted to set aside a dedicated two-hour block to ensure a thorough review. The taproot code is much cleaner and the previous one.

File: src/descriptor/tr.rs [hunk 1]
```diff
@@ -495,99 +494,84 @@ impl<Pk: FromStrKey> core::str::FromStr for Tr<Pk> {
...
         use crate::expression::{Parens, ParseTreeError};
 
-        expr_tree
-            .verify_toplevel("tr", 1..=2)
+        struct TreeStack<'s, Pk: MiniscriptKey> {
+            inner: Vec<(expression::TreeIterItem<'s>, TapTree<Pk>)>,
+        }
+
+        impl<'s, Pk: MiniscriptKey> TreeStack<'s, Pk> {
+            fn new() -> Self { Self { inner: Vec::with_capacity(128) } }
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply src/descriptor/tr.rs:510]: Yes, but we're talking about a savings of a kilobyte or so of heap space. We can maybe revisit this later as part of a "can we make this library work on constrained memory devices" project.

[the remark this answers is not in the document]
Comment (sanket1729) [reply src/descriptor/tr.rs:510]: Agreed. Good to merge this one PR

**Comment by apoelstra** [comment]
I will need to tweak the fuzztest commit with a workaround for https://github.com/nix-community/crate2nix/issues/373

Sorry for the delay -- it took me quite a while to track down that bug.

**Review by sanket1729 — APPROVED** [review]
utACK 6c8c34b347f9e3d531c9a0b771e6fa607a5c8e9b

**Review by apoelstra — COMMENTED** [review]
On 8235d5ffb669b6dae035dd04ebc9191b7b55fcc2 successfully ran local tests

**Pull request merged**